nemo/collections/nlp/modules/common/megatron/adapters/mcore_mixins.py

```diff
@@ -363,14 +371,51 @@ def glu(x):
 
         # LoRA logic
         if self.is_adapter_available():
-            lora_linear_fc2_adapter = self.get_adapter_module(AdapterName.LORA_4HtoH_ADAPTER)
-            if lora_linear_fc2_adapter and self.adapter_cfg[AdapterName.LORA_4HtoH_ADAPTER]['enabled']:
-                lora_output = lora_linear_fc2_adapter(intermediate_parallel)
-                output = output + lora_output
+            lora_fc2_adapter = self.get_adapter_module(AdapterName.LORA_4HtoH_ADAPTER)
+            lora_moe_fc2_adapter = self.get_adapter_module(AdapterName.LORA_MOE_4HtoH_ADAPTER)
+
+            lora_output = 0
+            if lora_fc2_adapter and self.adapter_cfg[AdapterName.LORA_4HtoH_ADAPTER]['enabled']:
+                lora_output = lora_fc2_adapter(intermediate_parallel)
+            elif lora_moe_fc2_adapter and self.adapter_cfg[AdapterName.LORA_MOE_4HtoH_ADAPTER]['enabled']:
+                lora_output = lora_moe_fc2_adapter(intermediate_parallel, expert_idx)
+
+            output = output + lora_output
 
         return output, output_bias
 
 
+class MCoreSequentialMLPMixin(SequentialMLP, MCoreAdapterModuleMixin):
+    def mcore_register_adapters(self):
+        """
+        We don't want the SequentialMLP layer to take any adapters. We only want to override the forward() behavior
+        """
+        pass
+
+    def forward(self, permuted_local_hidden_states, tokens_per_expert):
+        output_local = torch.zeros_like(permuted_local_hidden_states)
+        output_bias_local = None
+        if self.add_bias:
+            output_bias_local = torch.zeros_like(permuted_local_hidden_states)
+
+        cumsum_num_tokens = torch.cumsum(tokens_per_expert, dim=0)
+        # Insert zero at the begining for offset index's convenience
+        zero_tensor = torch.zeros(1, dtype=torch.long)
```

Comment:
@cuichenx do you know which device this uses cpu/gpu?

Reply:
this is copied from `SequentialMLP` (unfortunate limitation of this mcore mixin framework)

Reply:
mcore uses cumsum_num_tokens' device https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/transformer/moe/experts.py#L201

Reply:
ohh thanks for catching this, i must have copied from an older version of mcore!